### PR TITLE
[Enhancement] Tries to inject Vertx instance from CDI using a qualifier

### LINKED ISSUE #380: [Enhancement] Tries to inject Vertx instance from CDI using a qualifier
Currently the Vertx instance used in the smallrye opentelemetry comes from `Vertx.vertx()` which creates a new instance each time.

In some case, the Vertx instance is exposed with a CDI qualifier, it would be good if we can only inject the instance with the qualifier using CDI.

More information can be found at the Zulip discussion: https://wildfly.zulipchat.com/#narrow/stream/174184-wildfly-developers/topic/vertx.20subsystem.20integration

### PR DESCRIPTION
**Pull request by gaol**
Fixes #380 

This PR tries to add a new config `otel.exporter.vertx.cdi.identifier` to tries to inject the Vertx instance from the CDI with the qualifier, and it falls back to current way of `Vertx.vertx()` if there is no such configuration to keep compatible.

### CODE REVIEW

**Comment by jasondlee** [comment]
@gaol, This looks good. Once the formatting checks have been satisfied, I can merge this.

**Review by jasondlee — CHANGES_REQUESTED** [review]
@gaol, This looks good. Once the formatting checks have been satisfied, I can merge this.

**Comment by gaol** [comment]
@jasondlee from the discussion in https://github.com/smallrye/smallrye-reactive-messaging/pull/2772, do you see problem if using [Identifier](https://github.com/smallrye/smallrye-common/blob/2.6.0/annotation/src/main/java/io/smallrye/common/annotation/Identifier.java#L23) instead of [Named](https://github.com/jakartaee/inject/blob/2.0.1/src/main/java/jakarta/inject/Named.java#L38) qualifier in smallrye-opentelemetry ?  which would mean that there will be a dependency of `io.smallrye.common:smallrye-common-annotation` added to the exports artifact in this project, and a dependency of `io.smallrye.common.annotation` to the jboss module: `io.smallrye.opentelemetry` in WildFly.

**Comment by gaol** [comment]
@jasondlee I updated the PR to use `@io.smallrye.common.annotation.Identifier` from `smallrye-common-annotation`, would you please check again ? thank you :)

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (jasondlee) [inline implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:73]: What happens if there is no Vertx bean defined? Say someone sets that identifier property (or it comes in via some dependency), but no actual bean is defined in the system for some strange reason. This look up will fail, correct? Should the system fail to start, or should it fallback to the default (possibly with a WARN message)?

**Review by jasondlee — CHANGES_REQUESTED** [review]

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (gaol) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:73]: Yes, the look up will fail in that case. I am not sure if the system will fail to start, basically it will fail to `createExporter` I guess.

I think falling back to the default would be a better option. I think you mean `Vertx.vertx()` by `the default`, right ?

Comment (jasondlee) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:73]: That's what I had in mind, yes. I think NOT failing is probably the best option. It's certainly possible that the presence of that property with the absence of a bean is a bug, but it's certainly also possible that the entry is there unintentionally and not easily removed (e.g., third party jar), so I think a big, ugly warning and a reasonable fallback is probably the right choice.

**Comment by gaol** [comment]
Hello @jasondlee , I updated according to your feedback to log a warning message and return a new Vertx instance in case it cannot resolve from CDI. Would you please review again ? thank

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (Ladicek) [inline implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:41]: This should IMHO be called

```suggestion
    private static final String OTEL_EXPORTER_VERTX_CDI_QUALIFIER = "otel.exporter.vertx.cdi.identifier";
```

Because it is a value of the `@Identifier` qualifier, not a full qualifier (which would be very hard to denote).

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (Ladicek) [inline implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:82]:
```suggestion
                logger.log(Level.WARNING, "The Vertx instance with CDI qualifier @Identifier(\"{0}\") is not resolvable.",
```

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (Ladicek) [inline implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:86]: It is my personal opinion that we should try to resolve `Vertx` from CDI without a qualifier here. If that doesn't resolve, I'm not sure, but creating a new instance unconditionally doesn't seem like the worst idea.

**Comment by Ladicek** [comment]
Just interjected a few comments, feel free to ignore or do whatever you wish with them :-)

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (gaol) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:86]: I would prefer to just try to resolve it when a qualifier is specified, so it is on smallrye-opentelemetry's side to decide which one to use. If we try to resolve it from CDI without a qualifier, as long as there is a Vertx instance exposed in CDI container, it is forced to get injected to smallrye-opentelemtry.

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (gaol) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java]: Done

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (gaol) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java]: Done

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (Ladicek) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:86]: Sounds like exactly what I would expect :shrug:

Comment (jasondlee) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:86]: As it's written now, if there's an @Identifer defined but no bean properly annotated, we log a WARN and fall back to explicit Vertx instantation. If I understand you correctly, @Ladicek, you would like to put an attempt to find and unqualified Vertx bean in the middle before we fall back, correct? If that's so, then the logic would look like:

1. Check for qualified
2. Check for unqualified
3. Return local instance

That actually does seem like a safer and easier approach. *Requiring* the extra annotation, even in instances where it's not strictly needed, seems like extra work for no real value. The qualifier's intent is to allow the system to properly select Vertx beans in scenarios when there's more than one. If there's only one, then it's obvious which to select. If that bean is NOT what the user wants, then he or she should provide a properly annotated option.

**Review by jasondlee — CHANGES_REQUESTED** [review]
I left one last comment.

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (jasondlee) [inline implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:41]: One more thing: Since this is a SmallRye feature and not Otel, should this sysprop reflect that? Say, `smallrye.otel.exporter.vertx.cdi.identifier` or am I overthinking it?

Comment (gaol) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:41]: I am not sure, but would it break https://github.com/wildfly/wildfly/blob/34.0.0.Beta1/microprofile/telemetry-smallrye/cdi-provider/src/main/java/org/wildfly/extension/microprofile/telemetry/api/MicroProfileTelemetryCdiExtension.java#L39 if it starts with `smallrye` ?

**Comment by gaol** [comment]
@jasondlee I think exposing one instance of a bean using a qualifier can also leave it to the lookup side to decide if it wants the bean or not. Trying to lookup the bean without qualifier will hand over the decision out :)  I remembered in the zulip discussion that we want the lookup side be able to opt in, at least in the start time of integrating the vertx subsystem.

Please feel free to correct me if I am wrong. :)

File: implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java
Comment (jasondlee) [reply implementation/exporters/src/main/java/io/smallrye/opentelemetry/implementation/exporters/AbstractVertxExporterProvider.java:41]: Yeah, that's fair. Ignore this one. :)

**Comment by jasondlee** [comment]
> @jasondlee I think exposing one instance of a bean using a qualifier can also leave it to the lookup side to decide if it wants the bean or not. Trying to lookup the bean without qualifier will hand over the decision out :) I remembered in the zulip discussion that we want the lookup side be able to opt in, at least in the start time of integrating the vertx subsystem.
> 
> Please feel free to correct me if I am wrong. :)

I just reread that and you're correct. Given that, I think this looks great. I'll get it merged. Thanks!

**Review by jasondlee — APPROVED** [review]

**Comment by jasondlee** [comment]
HI, @gaol. One more thing: commits to this repo have to be signed. Once you've signed this commit, I'll merge it.

**Comment by gaol** [comment]
@jasondlee Updated to signed commit :)

**Comment by gaol** [comment]
@jasondlee sorry, I got the wrong idea, I understood that as `Signed-off-by` , will update again.

**Comment by gaol** [comment]
@jasondlee finally, I got the the signed commit :)

**Comment by jasondlee** [comment]
No worries. Got it. Thanks, @gaol , @Ladicek !

**Pull request merged**
